### ISSUE TITLE
sequence training does not work in the latest build as of March 3, 2016

### ISSUE THREAD

**Issue by hsiao06** [Microsoft/CNTK#203]
I successfully performed sequence training on a LSTM using a version of CNTK in January. However, when I reran the same experiment using the latest build (as of March 3), I ran into two problems:

1. in the latest build, numUttsPerMinibatch has to be set to 1 or it would blow up the GPU memory.
     - it was set to 40 in the original experiment
2. once I set numUttsPerMinibatch to 1, sequence training could run but the speed is 10 times slower (from 1500 frames per second to 150 frames per second). And after the first epoch, the model is corrupted. This is a speech recognition system and the word error rate goes from 40% to 90%.

Could anyone look into the issue? At the end of this post, I'm showing the configs and some of the logs in my two runs:

Thanks!

CNTK in January, sequence training on 1 GPU
 Epoch[ 1 of 4]-Minibatch[9301-9400]: SamplesSeen = 66377; TrainLossPerSample =  0.46002380; EvalErr[0]PerSample = 0.40625518; TotalTime = 31.8234s; SamplesPerSecond = 2085.8
 Epoch[ 1 of 4]-Minibatch[9401-9500]: SamplesSeen = 59149; TrainLossPerSample =  0.45716749; EvalErr[0]PerSample = 0.38670138; TotalTime = 42.0817s; SamplesPerSecond = 1405.6
 Epoch[ 1 of 4]-Minibatch[9501-9600]: SamplesSeen = 64309; TrainLossPerSample =  0.42722636; EvalErr[0]PerSample = 0.39994402; TotalTime = 31.9755s; SamplesPerSecond = 2011.2
 Epoch[ 1 of 4]-Minibatch[9601-9700]: SamplesSeen = 63357; TrainLossPerSample =  0.42443613; EvalErr[0]PerSample = 0.38887574; TotalTime = 40.9904s; SamplesPerSecond = 1545.7
 Epoch[ 1 of 4]-Minibatch[9701-9800]: SamplesSeen = 58138; TrainLossPerSample =  0.40836028; EvalErr[0]PerSample = 0.36391001; TotalTime = 41.1055s; SamplesPerSecond = 1414.4

CNTK as of March 3, sequence training on 1 GPU, numUttsPerMinibatch=1
 Epoch[ 1 of 4]-Minibatch[1401-1500]: SamplesSeen = 33025; TrainLossPerSample =  8.01973505; EvalErr[0]PerSample = 0.81798637; TotalTime = 152.5118s; SamplesPerSecond = 216.5
 Epoch[ 1 of 4]-Minibatch[1501-1600]: SamplesSeen = 28369; TrainLossPerSample =  7.93267299; EvalErr[0]PerSample = 0.82318728; TotalTime = 130.9632s; SamplesPerSecond = 216.6
 Epoch[ 1 of 4]-Minibatch[1601-1700]: SamplesSeen = 32906; TrainLossPerSample =  7.87816812; EvalErr[0]PerSample = 0.83692944; TotalTime = 153.1709s; SamplesPerSecond = 214.8
 Epoch[ 1 of 4]-Minibatch[1701-1800]: SamplesSeen = 30568; TrainLossPerSample =  7.79491625; EvalErr[0]PerSample = 0.78850432; TotalTime = 143.0214s; SamplesPerSecond = 213.7
 Epoch[ 1 of 4]-Minibatch[1801-1900]: SamplesSeen = 36435; TrainLossPerSample =  7.61244682; EvalErr[0]PerSample = 0.76316728; TotalTime = 168.5101s; SamplesPerSecond = 216.2

my config:

NdlDir=$ExpDir$/configs
command=$command$
numCPUThreads=$numThreads$
precision=float
stderr=$ExpDir$/log/$logFile$

shareNodeValueMatrices=true

parallelTrain=false

EditModel=[
    action=edit
    oldModel=$origModelName$
    newModel=$ExpDir$/$modelName$.0
    scaleFactor=$smoothFactor$
    labelDim=$labelDim$
    editPath=$NdlDir$/edit.mel
]

TrainModel=[
    action=train
    modelPath=$ExpDir$/$modelName$

    # deviceId=-1 for CPU, >=0 for GPU devices
    deviceId=$DeviceNumber$
    traceLevel=1
    useValidation=true

    NDLNetworkBuilder=[
        ndlMacros=$NdlDir$/default_macros.ndl
        networkDescription=$NdlDir$/model.ndl
    ]

    SGD=[
        epochSize=0
        minibatchSize=$minibatchSize$
        learningRatesPerSample=$lratePerSample$
        momentumPerMB=$momentum$
        L2RegWeight=$l2RegWeight$
        dropoutRate=$dropoutRate$
        maxEpochs=$maxEpochs$
        numMBsToShowResult=$evaluatePeriod$

        #parallelTrain = [
        #    parallelizationMethod = "DataParallelSGD"
        #    distributedMBReading = true
            # Additional optional parameters are: useZeroThresholdFor1BitQuantization
        #    dataParallelSGD = [
        #        gradientBits = 1
        #    ]
        #]

        #settings for Auto Adjust Learning Rate
        AutoAdjust=[
            reduceLearnRateIfImproveLessThan=0
            loadBestModel=true
            increaseLearnRateIfImproveMoreThan=1000000000
            learnRateDecreaseFactor=0.5
            autoAdjustLR=AdjustAfterEpoch
            learnRateAdjustInterval=1
        ]

        gradientClippingWithTruncation=true
        clippingThresholdPerSample=$clippingThresholdPerSample$

    ]

    # Parameter values for the reader
    reader=[
        # reader to use
        readerType=Kaldi2Reader
        readMethod=blockRandomize
        frameMode=false
        Truncated=$truncated$
        nbruttsineachrecurrentiter=$numUttsPerMinibatch$
        miniBatchMode=Partial
        randomize=Auto
        verbosity=1
        seqTrainCriterion=$criterion$
        features=[
            dim=$featDim$
            scpFile=$inputCounts$
            rx=$inputFeats$
            featureTransform=$featureTransform$
        ]

        labels=[
            mlfFile=$inputLabels$
            labelDim=$labelDim$
            labelMappingFile=$labelMapping$
        ]

        derivatives=[
            type=seqTrainDeriv
        ]

        objectives=[
            type=seqTrainObj
        ]

        denlats=[
            rx=$inputDenLats$
            silPhoneList=$silPhoneList$
            kaldiModel=$kaldiModel$
            oldAcousticScale=0.0
            acousticScale=$acwt$
            lmScale=1
            oneSilenceClass=$oneSilenceClass$
        ]

        alignments=[
            rx=$inputAli$
        ]
    ]
]

**Comment by yzhang87** [comment]
I'll look into this but if anyone has any hint please let me know.

**Comment by yzhang87** [comment]
I think this is not the real problem. This is because you didn't redo the model editing. If you finish the model editing, the bug will be like post #202 .

**Comment by chenguoguo** [comment]
By model editing, Yu meant you have to create the your initial sequence model from your CE model again. We are looking into issue #202 right now.

**Comment by hsiao06** [comment]
Yu Zhang is right. The model editing job failed which I didn't notice. After fixing that, it will trigger the bug in #202. Thanks

**Comment by yzhang87** [comment]
@hsiao06 just send a pull request. You can sync with my fork if you need it now.
